### ISSUE TITLE
Cache disabled after run "composer update" command

### ISSUE THREAD

**Issue by ghost** [magento/magento2#17634]
@magento-engcom-team give me 2.2.3 instance

I'm not sure if that is an expected behavior but, after trying to execute:

`composer update`

all caches are being disabled.

Happened every time I had to run composer update since I'm working with Magento 2. First time it took me around 1 complete day debugging. 
Caches being disabled was the last thing I thought of when I did not enter **"bin/magento cache:disable"**

### Preconditions
Magento 2.2.3
PHP 7.0.31
nginx 1.10.3

### Steps to reproduce

$ composer update
$ php bin/magento cache:status

```
Current status:
                        config: 0
                        layout: 0
                    block_html: 0
                   collections: 0
                    reflection: 0
                        db_ddl: 0
                  social_feeds: 0
                           eav: 0
         customer_notification: 0
            config_integration: 0
        config_integration_api: 0
                     full_page: 0
                     translate: 0
             config_webservice: 0

```
### Expected results
composer update packages/dependencies without disabling Magento 2 cache.

### Actual results
composer update packages/dependencies and then is disabling Magento 2 cache.

**Comment by magento-engcom-team** [comment]
Hi @bruno-serfe. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation.

@bruno-serfe do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by ghost** [comment]
@bruno-serfe , thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by hostep** [comment]
@engcom-backlog-andrii or @bruno-serfe, can you guys add some more exact steps to reproduce?

I've seen this happening occasionally that caches get disabled somehow by "something", but it is certainly not for **every** `composer update` call, there must be some other condition, otherwise I would run into this problem multiple times a day, which simply isn't happening.

Exact conditions and steps to reproduce would be very much appreciated!

**Comment by ghost** [comment]
Hi @hostep I've tried in my work environment and it's just that, I've bring the latest changes from the git repo and then executed "composer update" just "as is" with no other previous command and is just disabling entire magento 2 cache, commands that are executed after composer update are:

```
$ composer install
$ php bin/magento setup:upgrade
$ php bin/magento cache:flush
$ php bin/magento deploy:mode:set production
```

and that's all., hope that helps. Magento version is 2.2.3

Kind regards,
Bruno

**Comment by hostep** [comment]
Thanks @bruno-serfe for the quick reply!
It's really strange that this happens. Is there something which manipulates the `app/etc/env.php` file in your steps? Maybe some composer plugin? Because the `env.php` file keeps track of if a cache is enabled or disabled. So "something" must be changing that file for some reason.

I'm particularly interested in @engcom-backlog-andrii's reply, because he will most likely have tested this on a clean installation, so I'm curious to what his exact steps were...

**Comment by ghost** [comment]
Hi @hostep 
Regarding: 

> Is there something which manipulates the app/etc/env.php file in your steps?

I really don't know, but I'll consider to check that, if it helps to resolve this.
I'll post back when I've some update about that.

Thanks to you guys.

**Comment by ghost** [comment]
Hi @hostep , i can reproduce this issue on fresh installiation with steps on this issue. 
Preconditions
 I have already installing clean magento instance 2.2-dev or 2.3-dev on  locall
Steps: 
1. Run php bin/magento deploy:mode:set production
2. Run command composer update for check updates for packeches
3. Run command php bin/magento cache:status

**Comment by hostep** [comment]
Thanks for the update @engcom-backlog-andrii 

Unfortunately I still can't reproduce:
- using PHP 7.0.31 & composer 1.7.1
- ran `composer create-project ...` to setup Magento OS 2.2.5
- ran `bin/magento setup:install ...`
- ran `bin/magento deploy:mode:set production`
- ran `composer update`
- ran `bin/magento cache:status`

Result: still all caches are enabled.

Really strange, so there must be some kind of environment setting which is different then mine. Are you guys using Redis as cache storage maybe? I'm using the filesystem as storage in case that matters.

Anyway, since @engcom-backlog-andrii can reproduce, that means that Magento has a way of reproducing which should make it easier to try to debug this, hopefully they'll be able to figure out what goes wrong then here.

**Comment by ghost** [comment]
Hi @hostep , Yes I'm using Redis in my environment, maybe you should try with Redis.

**Comment by ghost** [comment]
Just in case, my composer version is: 

> Composer version 1.6.3 2018-01-31 16:28:17

**Comment by ghost** [comment]
Hi guys! Is there any update about this? thanks!

**Comment by lewisvoncken** [comment]
Hi @bruno-serfe , @hostep 

It looks like that this problem persists when the var/.regenerate flag is set. This file will be created when enabling a module through the command line.

@bruno-serfe can you please confirm this?

**Comment by yaroslav-zenin** [comment]
also reproduced for me on 2.2.3 after `composer update`, cache engine Redis

**Comment by PascalBrouwers** [comment]
I have this too sometimes. But then with developer mode and no Redis. Usually when running setup:upgrade after enabling a module. Can't reproduce unfortunately.

**Comment by WesleyVestjens** [comment]
Does anyone have any more info about this? We're running into this issue on a regular basis and it's becoming a real annoying issue...

**Comment by Vormillion** [comment]
In our case this issue is happening randomly without any activity on server. At the end of deployment script I am enabling Magento cache (just in case...) and printing cache status. Sometime after few minutes cache is automatically off, but there is no pattern.

**Comment by cscd98** [comment]
I've seen this in our development environment before. It's rather annoying.

**Comment by nickpiro** [comment]
This is still happening for us as well. Not sure where it is coming from, but I just ran a composer command to update our Algolia module "composer update algolia/algoliasearch-magento-2" and it disabled cache. This also happens randomly when we run setup:upgrade as well.

**Comment by Zyles** [comment]
Yep discovered this today after 2 hours of "WTF is going on" with performance.

**Comment by rafaelstz** [comment]
I'm having this issue in some projects also. Try to do these steps:

- Remove `pub/media/catalog/product/cache` . (Removing this folder frees up space.)

- Run `bin/magento catalog:image:resize` to generate a new image cache. (This step is necessary because we’ve changed the path to cached images and must remove the previously cached images.)

**Comment by tzyganu** [comment]
@engcom-backlog-tomash Can you explain in a few words what causes this? We might want to fix our instances before the fix is released. 
Thanks in advance.

**Comment by onepack** [comment]
I can acknowledge that it is a standard procedure now for me as well to check the caches after composer update.
They will become disabled and performance will drop dramatically when all crons are running as well.

**Comment by bartoszkubicki** [comment]
Maybe it is something with all commands objects creation while running any command?

**Comment by melaxon** [comment]
Sometimes the same happens also after setup:upgrade on developer mode and no redis on 2.2.x and 2.3.1 on both local and remote host. But I can not reproduce it intentionally. No idea what it depends on

**Comment by adrian-martinez-interactiv4** [comment]
For anyone interested, I think I can bring some light about this issue. It is a **concurrency (race condition) problem in \Magento\Framework\Code\GeneratedFiles::cleanGeneratedFiles, when the var/.regenerate flag is set, and more than one process/request tries to clean generated files**.

It is more likely to happen when you have cron enabled, and many cron groups using use_separate_process config. When more than one process are trying to clean the same, FileIterator fails with different messages similar to: `FilesystemIterator::__construct(/Users/adrianmartinez/Sites/r2-project-develop-b2b/environments/2-2-develop-b2b/magento/generated/code): failed to open dir: No such file or directory.`

Moving up call to `$this->write->delete(self::REGENERATE_FLAG);`, just after flag existence check resolves the problem, as the first arriving process marks itself as responsible of cleaning files.

I leave here a **demo video about how to replicate the problem**:
https://youtu.be/9-X1cIIY7y8

And the script used for it:
```
#!/bin/bash

# \Magento\Framework\Code\GeneratedFiles has a concurrency problem
# Create regenerate flag and launch parallel commands that try to regenerate at the same time
# This is a real case, cron:run launches stand alone processes in parallel

# Created by magento composer installer upon code install or module enable/disable
touch var/.regenerate

# Launch parallel commands
# Error differs each execution, sometimes it even works
bin/magento cron:run --group=ddg_automation --bootstrap=standaloneProcessStarted=1 2>&1 &
bin/magento cron:run --group=index --bootstrap=standaloneProcessStarted=1 2>&1 &

wait
echo "All done"
```

**Comment by VZeroCool** [comment]
This issue still present in 2.3.1. It getting disabled randomly. Does any one has fix ?

**Comment by lucscu** [comment]
@adrian-martinez-interactiv4 very great hint, thx you.
I can't test it, maybe someone could verify your answer.

**Comment by reinaldomendes** [comment]
It's  a race condition Magento\Framework\Code\GeneratedFiles::cleanGeneratedFiles.

You can simulate the problem using the code bellow.



```php
<?php
// index-test.php
/**
 * Application entry point
 *
 * Example - run a particular store or website:
 * --------------------------------------------
 * require __DIR__ . '/app/bootstrap.php';
 * $params = $_SERVER;
 * $params[\Magento\Store\Model\StoreManager::PARAM_RUN_CODE] = 'website2';
 * $params[\Magento\Store\Model\StoreManager::PARAM_RUN_TYPE] = 'website';
 * $bootstrap = \Magento\Framework\App\Bootstrap::create(BP, $params);
 * \/** @var \Magento\Framework\App\Http $app *\/
 * $app = $bootstrap->createApplication(\Magento\Framework\App\Http::class);
 * $bootstrap->run($app);
 * --------------------------------------------
 *
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

try {
    require __DIR__ . '/app/bootstrap.php';
} catch (\Exception $e) {
    echo <<<HTML
<div style="font:12px/1.35em arial, helvetica, sans-serif;">
    <div style="margin:0 0 25px 0; border-bottom:1px solid #ccc;">
        <h3 style="margin:0;font-size:1.7em;font-weight:normal;text-transform:none;text-align:left;color:#2f2f2f;">
        Autoload error</h3>
    </div>
    <p>{$e->getMessage()}</p>
</div>
HTML;
    exit(1);
}

$bootstrap = \Magento\Framework\App\Bootstrap::create(BP, $_SERVER);
/** @var \Magento\Framework\App\Http $app */
//$app = $bootstrap->createApplication(\Magento\Framework\App\Http::class);

$di = $bootstrap->getObjectManager();
/*@var $di Magento\Framework\App\ObjectManager*/


$handler = $di->get(Magento\Framework\App\ErrorHandler::class);
set_error_handler([$handler, 'handler']);



/**
 *
 * @param \Magento\Framework\ObjectManagerInterface $di
 * @param null|string $code [frontend, adminhtml]
 */
function __mage_init__( \Magento\Framework\ObjectManagerInterface $di, $code=null){
    $state = $di->get(Magento\Framework\App\State::class);
    $configLoader = $di->get(\Magento\Framework\ObjectManager\ConfigLoaderInterface::class);
    $areaList = $di->get(Magento\Framework\App\AreaList::class);
    $request = $di->get(Magento\Framework\App\Request\Http::class);

    $areaCode = $code ? $code : $areaList->getCodeByFrontName($request->getFrontName());
    $state->setAreaCode($areaCode);
    $di->configure($configLoader->load($areaCode));
}

// $code = 'frontend';
// $code = 'adminhtml';
$code = null;
__mage_init__($di, $code);


$generatedFiles = $di->get(Magento\Framework\Code\GeneratedFiles::class);
$generatedFiles->cleanGeneratedFiles();
```

```bash
   touch var/.regenerate; php -f index-test.php & php -f index-test.php
```

**Comment by chickenland** [comment]
This issue is recurring for us, so looking for a solution.  The information from @adrian-martinez-interactiv4 - I'd suggest moving this up, but I feel like it should lock somehow until it completes, rather than carrying on with other requests - which might lead to files that are generated and cached?

**Comment by dandelionred** [comment]
Just got this cache disabled mystery in 2.3.1. Thanks for the fix

**Comment by ladle3000** [comment]
I see @adrian-martinez-interactiv4  has identified it, but I'm unclear what the 'fix' is. Can someone please put the fix plainly for others that land here too?

**Comment by itecedor** [comment]
I am also experiencing this problem intermittently in 2.3.2, but without a `var/.regenerate` file at all. It just happened again in production even though I had not run composer, and seems to happen consistently given enough time.

Based on the commends in this issue I suspect it has to do with concurrent cron jobs -- I have crons configured to run 3 processes at once. Will test bringing that down to 1 cron job to see if it helps.

**Comment by elfeffe** [comment]
@itecedor Any news?
I have the same problem

**Comment by ci2014** [comment]
Same here

**Comment by lewisvoncken** [comment]
I think I have fix. I will create a PR asap

**Comment by lewisvoncken** [comment]
Thanks to @adrian-martinez-interactiv4  have created a PR
https://github.com/magento/magento2/pull/24560

**Comment by magento-engcom-team** [comment]
Hi @bruno-serfe. Thank you for your report. 
The issue has been fixed in magento/magento2#24892 by @adrian-martinez-interactiv4 in 2.3-develop branch
Related commit(s):
 - [af9ca675f2102aacf4003afa4d8fa779cf0b5e4e](https://github.com/magento/magento2/commit/af9ca675f2102aacf4003afa4d8fa779cf0b5e4e)

The fix will be available with the upcoming 2.3.4 release.
